tests/console/yast2_lan_hostname_dhcp_no.pm

```diff
@@ -0,0 +1,37 @@
+# SUSE's openQA tests
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Package: yast2-network
+# Summary: Verify that correct value is stored in network config when
+# setting hostname via DHCP to 'no' in YaST2 lan module (https://bugzilla.suse.com/show_bug.cgi?id=984890)
+# Maintainer: QA SLE YaST team <qa-sle-yast@suse.de>
+
+use parent 'yast2_lan_hostname_base';
+use strict;
+use warnings;
+use testapi;
+use scheduler qw(get_test_suite_data);
+use YaST::Module;
+
+sub run {
+    my $test_data = get_test_suite_data();
+    YaST::Module::open({module => 'lan', ui => $test_data->{ui}});
+    my $network_settings = $testapi::distri->get_network_settings();
+    $network_settings->set_hostname_via_dhcp({
+            dhcp_option     => 'no',
```

Comment:
why don't just represent those changes in test data instead of creating new tests? You could iterate in one test over those different configurations. Most modules should do that, open/change/apply/check & repeat over a list of same hash struct of test data.

Reply:
This is what I wanted to discuss and I briefly described the reason for this during our meeting.

- The initial reason that gave me such idea was because we have different validation for different cases. (e.g. for eth we validate with 3 scripts, for no - with one script). It is technical thing and potentially I can solve it, for example by storing those scripts in test data.

- The second reason is that if store them in one test module and just iterate, then if one of them will fail, others will not be executed at all, even though they may pass of have their own issues. 
   For example, during implementation of these tests I noticed that if 'no' is selected (which is default), then network configuration is not changed and services are not restarted. But in case of 'yes', the test become slower, because configuration save started, then network services restart, firewall restarts. Test may catch some issues. But if the case with 'no' was failed before in a loop, we would not see those problems.

- And last, but probably the main reason is conceptual:) It is that I see them as 3 separate test cases.

   E.g. we call `add_partition` several times in a loop in partitioner test, because we need to create several partitions that are needed for the test, so that we make setup for validation. But it is just one test case: configure partitions and validate the system is installed with this setup.

   In case of yast2 lan settings - it is three different test cases. We can remove any of them from the test suite, change their order and add more test cases for yast2 lan. And after we apply each change, we validate that the change was stored. One setup - one verification.

I hope it helped to clarify my idea. If you still have some questions, doubts or ideas, feel free to contact me via chat tool.

Reply:
> This is what I wanted to discuss and I briefly described the reason for this during our meeting.
> 
>     * The initial reason that gave me such idea was because we have different validation for different cases. (e.g. for eth we validate with 3 scripts, for no - with one script). It is technical thing and potentially I can solve it, for example by storing those scripts in test data.
> 

those script can still be in the test, you just run them conditionally depending on your test data. IMO it is an overkill to have a test for a 'no' thing and a 'yes' thing.

>     * The second reason is that if store them in one test module and just iterate, then if one of them will fail, others will not be executed at all, even though they may pass of have their own issues.
>       For example, during implementation of these tests I noticed that if 'no' is selected (which is default), then network configuration is not changed and services are not restarted. But in case of 'yes', the test become slower, because configuration save started, then network services restart, firewall restarts. Test may catch some issues. But if the test case with 'no' was failed before in a loop, we would not see those problems.
> 

you could accumulate the errors, I don't see a problem with that if it is that important.

>     * And last, but probably the main reason is conceptual:) It is that I see them as 3 separate test cases.
>       E.g. we call `add_partition` several times in a loop in partitioner test, because we need to create several partitions that are needed for the test, so that we make setup for validation. But it is just one test case: configure partitions and validate the system is installed with this setup.

They are three test or steps (doesn't matter how we call it), but you can still have one code that work those three cases with three different set of test data.

>       In case of yast2 lan settings - it is three different test cases. We can remove any of them from the test suite, change their order and add more test cases for yast2 lan. And after we apply each change, we validate that the change was stored. One setup - one verification.
> 

I would like to be able to remove one of them just touching test data.

> 
> I hope it helped to clarify my idea. If you still have some questions, doubts or ideas, feel free to contact me via chat tool.

Reply:
but this is just my view here, that for simple conditions, based on presence/not presence of test data, the test is still readable, which I think is the case.

Reply:
I would say that test data approach looks natural partly because in the installation tests we cannot validate things straight away. In the installed system, we have opportunity to define everything in single place, which can be test module and not yaml. I thought of having single module which tests all 3 dhcp option, but we also can just group them once we have better visualization in openQA. This approach allows us to test all 3 options still, even one of them fails without extra hassle to revert changes. We still should extract common parts (there is not much in this case). So I tend to upvote current proposal to abstract data driven test, at least in this particular case.

Reply:
2 vs 1 and I also see the benefits of this solution. 
I will try to explore how the code looks with intermediate granularity (all in one when possible and simple) in my next PRs
Resolving this conversation.